packages/angular/projects/angular/src/internal/dynamic-template/dynamic-template.component.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ComponentRef,
+  EmbeddedViewRef,
+  Input,
+  TemplateRef,
+  ViewContainerRef,
+} from '@angular/core';
+import { UnknownRecord } from '@devextreme/core';
+import { AngularTemplate } from '../types';
+import { ViewComponent } from './view.component';
+
+interface AngularTemplateContext<T> {
+  $implicit: T;
+}
+
+@Component({
+  selector: 'dx-dynamic-template',
+  template: '',
+  styles: [':host {display: none; }'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class DynamicTemplateComponent<
+  TComponent extends ViewComponent,
+  TData extends UnknownRecord,
+  > {
+  @Input() set template(template: AngularTemplate<TComponent>) {
+    if (template === this.templateValue) {
+      return;
+    }
+
+    this.setTemplate(template);
+  }
+
+  @Input() set data(data: TData) {
+    if (data === this.dataValue) {
+      return;
+    }
+
+    this.setTemplateData(data);
+  }
+
+  private templateValue?: AngularTemplate<TComponent>;
+
+  private dataValue?: TData;
+
+  private componentRef?: ComponentRef<TComponent>;
+
+  private templateRef?: EmbeddedViewRef<AngularTemplateContext<TData>>;
+
+  constructor(
+    private viewContainerRef: ViewContainerRef,
+  ) {
+  }
+
+  private setTemplate(template: AngularTemplate<TComponent>): void {
+    if (!template) {
+      return;
+    }
+
+    this.templateValue = template;
+
+    if (template instanceof TemplateRef) {
+      this.templateRef = this.viewContainerRef.createEmbeddedView(template);
+    } else {
+      this.componentRef = this.viewContainerRef.createComponent(template);
+    }
+
+    this.setTemplateData(this.dataValue);
+  }
+
+  private setTemplateData(data?: TData): void {
+    if (!data) {
+      return;
+    }
+
+    this.dataValue = data;
+
+    if (this.componentRef) {
+      this.setDataToComponent();
+    } else if (this.templateRef) {
+      this.templateRef.context = { $implicit: this.dataValue! };
+    }
+  }
+
+  private setDataToComponent(): void {
+    const { instance } = this.componentRef!;
+
+    (Object.keys(this.dataValue!) as (keyof TData)[])
+      // TODO: Think about optimization here.
+      // .filter((key) => Object.prototype.hasOwnProperty.call(instance, key))
+      .forEach((key) => {
```

Comment:
Minor: Alternative for forEach here is for ... of.

Reply:
I think you mean ```for ... in ``` not ``` for ... of```

If we use ```for ... in``` here, it will look like this:
```ts
    for (const key in this.dataValue!) {
      // It's for the eslint rule 'guard-for-in'
      if (Object.prototype.hasOwnProperty.call(instance, key)) {
        (instance as UnknownRecord)[key] = this.dataValue![key];
      }
    }
```    

But the ```hasOwnProperty``` check is not suitable for us, because the view component may have the ```@Input()``` field in ts, but will not have it in runtime.

E.g. components like this
```ts
@Component({...})
export class MyComponent {
  @Input() myField?: string;
}
```
in runtime will not have ```myField``` until it will be initialized :)

If we implement your suggestion, we will need to disable the eslint rule 'guard-for-in' :(
So, let's leave it as is

Reply:
Yes. It's my mistake. Object.entries or Object.getOwnPropertyNames is more sutable suggestion for here.